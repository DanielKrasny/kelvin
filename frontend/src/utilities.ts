type Method = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';

/**
 * Get data from API
 * @param url url to fetch
 * @param data data which will be sent to the API
 * @param method HTTP method, if you want to override the default GET/POST
 * @note If data is passed, the request will be a POST request, otherwise a GET request, if not overridden by method parameter
 * @returns $ReturnType, if the request was successful, otherwise undefined
 */
export const getFromAPI = async <$ReturnType>(
    url: string,
    data?: unknown,
    method?: Method
): Promise<$ReturnType | undefined> => {
    try {
        const response = await fetch(url, {
            method: method || (data ? 'POST' : 'GET'),
            body: data ? JSON.stringify(data) : undefined
        });

        if (!response.ok) {
            return undefined;
        }

        const json = await response.json();
        return json as $ReturnType;
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (_) {
        return undefined;
    }
};

/**
 * Generate array with range of numbers from start
 * @param size size of the array
 * @param start starting number
 * @returns array of numbers from start to start + size
 */
export const generateRange = (size: number, start = 0) => {
    return Array.from({ length: size }).map((_, i) => i + start);
};

/**
 * Generate string date format from date
 * @param date Date object or string in format 'YYYY-MM-DD' or 'YYYY-MM-DDTHH:MMZ'
 * @param withTime Include time in the output
 * @returns If withTime is false 'YYYY-MM-DD', otherwise 'YYYY-MM-DD HH:MM'
 */
export const getDate = (date: Date | string, withTime = false) => {
    let d: Date;
    if (typeof date === 'string') {
        d = new Date(date);
    } else {
        d = date;
    }

    const year = d.getFullYear();
    const month = (d.getMonth() + 1).toString().padStart(2, '0');
    const day = d.getDate().toString().padStart(2, '0');

    if (!withTime) {
        return `${year}-${month}-${day}`;
    }

    const hours = d.getHours().toString().padStart(2, '0');
    const minutes = d.getMinutes().toString().padStart(2, '0');
    return `${year}-${month}-${day} ${hours}:${minutes}`;
};
