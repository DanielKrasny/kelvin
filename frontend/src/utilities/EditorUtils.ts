import CodeMirror from 'codemirror';

// Remove string Type Union from EditorConfiguration
export type ExtraKeys<$BaseType> = $BaseType extends string ? never : $BaseType;

export type Arrayable<$Type> = $Type | $Type[];

export const inArrayable = <T>(arrayable: Arrayable<T>, value: T) => {
    if (Array.isArray(arrayable)) {
        return arrayable.some((item) => item === value);
    }
    return arrayable === value;
};

export const arrayify = <T>(arrayable: Arrayable<T>) => {
    if (Array.isArray(arrayable)) return arrayable;
    return [arrayable];
};

/**
 * Type definition for extension
 * Each extension is a function. This function can accept these parameters:
 * Extension function is the main code for extension. You can accept these parameters:
 * - fileName - current file filename
 * - code - current code in editor
 * - editor - current editor instance
 * - type - gives you information about context, in which was function called. It can be one of these:
 *   - setup - This is called when editor is created (first time), or filename changed
 *   - lint - This is called when linting is requested (after code change)
 *   - hint - This is called when hinting is requested (after user presses ctrl+space)
 *
 * From this function you can return these things:
 * - Undefined for case, where you don't care about for example current file/extension (extension
 *   for json file doesn't do anything with python file, etc..), empty object is also valid return value
 *   and do same thing as undefined
 * - Object containing these keys:
 *   - spellCheck - Controlls if spellcheck should be enabled/disabled (only for setup)
 *   - gutters - List of gutters to set for current editor (only for setup)
 *   - hint - This will be array of hints, for user to select them
 *   - lint - Lint warning/errors for current code
 * - You don't need to set every key in object, only these, you do care. For example:
 *   - if you only care for hints, return something like this:
 *     ```TS
 *     return {
 *        hint: {
 *            list: ["const", "console"],
 *            from: {
 *               ch: 2,
 *               line: 2
 *            },
 *            to: {
 *                ch: 5,
 *                line: 2
 *            }
 *        }
 *     }
 *     ```
 *   - or if you want to enable spellcheck on setup:
 *     ```TS
 *     return {
 *        spellCheck: true
 *     }
 *     ```
 */
export type EditorExtension = (
    fileName: string,
    code: string,
    editor: CodeMirror.Editor,
    type: 'setup' | 'lint' | 'hint'
) =>
    | undefined
    | {
          spellCheck?: boolean;
          gutters?: string[];
          hint?: CodeMirror.Hint;
          lint?: CodeMirror.LintError[];
      };

export const getExtension = (filename: string) => {
    if (!filename) return;

    const parts = filename.split('.');
    return parts[parts.length - 1].toLowerCase();
};
