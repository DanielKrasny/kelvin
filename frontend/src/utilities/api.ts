export function csrfToken() {
    return document.querySelector('meta[name=csrf-token]').getAttribute('content');
}
