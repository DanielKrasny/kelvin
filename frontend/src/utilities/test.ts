import { vi } from 'vitest';

/**
 * Mock fetch requests
 * Currently accept one parameter, which is object, mapping url to
 * response object, which consists of okStatus and return data
 * returned by json() method
 */
export const mockFetch = (
    urlsMap: Record<
        string,
        {
            okStatus: boolean;
            data: unknown;
        }
    >
) => {
    const keys = Object.keys(urlsMap);

    vi.stubGlobal(
        'fetch',
        vi.fn((url: string) => {
            if (!keys.includes(url)) {
                return Promise.resolve({
                    ok: false,
                    json: () => Promise.resolve(undefined)
                });
            }

            const resultData = urlsMap[url];

            return Promise.resolve({
                ok: resultData.okStatus,
                json: () => Promise.resolve(resultData.data)
            });
        })
    );
};
