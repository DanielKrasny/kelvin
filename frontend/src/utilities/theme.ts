import { effect, ref } from 'vue';
import { localStorageStore } from './storage';

export type ThemeValue = 'auto' | 'light' | 'dark';
export const currentTheme = localStorageStore<ThemeValue>('theme', 'auto');

//this ref should be used for the current set theme in other parts of application
export const theme = ref<'light' | 'dark'>('light');

//function which detects current theme from media query
const htmlData = document.querySelector('html').dataset;
const themeMedia = window.matchMedia('(prefers-color-scheme: light)');

function detectTheme() {
    theme.value = themeMedia.matches ? 'light' : 'dark';
}

//since we want to have listener on media query to update theme
//every time when user changes theme in browser, we can't use
//computed, and we need to use effect
effect(() => {
    if (currentTheme.value === 'auto') {
        themeMedia.addEventListener('change', detectTheme);
        detectTheme();
        return;
    }
    //in future if this will became SPA, and ColorTheme component will not
    //be presented on every page we should call this in onUnmounted
    themeMedia.removeEventListener('change', detectTheme);

    theme.value = currentTheme.value;
});

//every time theme changes we want to update dataset in
//html tag for bootstrap to change theme
effect(() => {
    htmlData.bsTheme = theme.value;
});
