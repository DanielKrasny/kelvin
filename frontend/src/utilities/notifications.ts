import { getDataWithCSRF } from './api';

//@TODO: complete null unions if missing
export type Notification = {
    id: number;
    level: 'info'; //@TODO: add other levels
    recepient: number;
    unread: boolean;
    actor_content_type: number;
    actor_object_id: string;
    verb: string;
    description: string | null;
    timestamp: string;
    public: boolean;
    deleted: boolean;
    emailed: boolean;
    important: boolean;
    actor: string;
    action_object: string;
    action_object_url: string;
    custom_text?: string;
    target?: string;
};

export const getNotifications = async () => {
    const data = await getDataWithCSRF<{ notifications: Notification[] }>('/notification/all');
    return data.notifications;
};

export const markRead = async (id: number) => {
    const data = await getDataWithCSRF<{ notifications: Notification[] }>(
        '/notification/mark_as_read/' + id,
        'POST'
    );
    return data.notifications;
};

export const markAllRead = async () => {
    const data = await getDataWithCSRF<{ notifications: Notification[] }>(
        '/notification/mark_as_read',
        'POST'
    );
    return data.notifications;
};

const getPublicKey = () => {
    return document.querySelector<HTMLMetaElement>('meta[name="django-webpush-vapid-key"]').content;
};

const urlB64ToUint8Array = (base64String: string) => {
    const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
    const base64 = (base64String + padding).replace(/-/g, '+').replace(/_/g, '/');

    const rawData = window.atob(base64);
    const outputArray = new Uint8Array(rawData.length);

    for (let i = 0; i < rawData.length; ++i) {
        outputArray[i] = rawData.charCodeAt(i);
    }
    return outputArray;
};

const getSubscription = async (reg: ServiceWorkerRegistration) => {
    const subscribeOpts = {
        userVisibleOnly: true,
        applicationServerKey: urlB64ToUint8Array(getPublicKey())
    };

    try {
        let sub = await reg.pushManager.getSubscription();
        if (sub) {
            return sub;
        }

        sub = await reg.pushManager.subscribe(subscribeOpts);

        const browser = navigator.userAgent
            .match(/(firefox|msie|chrome|safari|trident)/gi)[0]
            .toLowerCase();
        const data = {
            status_type: 'subscribe',
            subscription: sub.toJSON(),
            browser: browser,
            group: null
        };

        await getDataWithCSRF('/webpush/save_information', 'POST', data, {
            'Content-Type': 'application/json'
        });
        return sub;
    } catch (exception) {
        console.log(exception);
    }

    return null;
};

export const subscribePushNotifications = async (reg: ServiceWorkerRegistration) => {
    const isEnabled = (await getSubscription(reg)) !== null;
    return isEnabled;
};

export type PushNotifications =
    | {
          supported: true;
          enabled: boolean;
          reg: ServiceWorkerRegistration;
      }
    | {
          supported: false;
          enabled: null;
          reg: null;
      };

export const getPushStatus = async (): Promise<PushNotifications> => {
    let reg: null | ServiceWorkerRegistration = null;
    if ('serviceWorker' in navigator && 'PushManager' in window && getPublicKey()) {
        try {
            reg = await navigator.serviceWorker.register('/static/service-worker.js');
            if (!reg.showNotification) {
                return;
            }
            return {
                supported: true,
                enabled: await subscribePushNotifications(reg),
                reg
            };
        } catch (err) {
            console.log(err);
            return {
                supported: true,
                enabled: false,
                reg
            };
        }
    }

    return {
        supported: false,
        enabled: null,
        reg: null
    };
};
