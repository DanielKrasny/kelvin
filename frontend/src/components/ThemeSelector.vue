<script lang="ts" setup>
/**
 * This component is used to select theme used on website.
 * After opening the dropdown, it shows the list of avalable color themes.
 * Upon chosing the theme, it will update binded value to this component
 * It is available to both teachers and students, and it is accesible from
 * main layout on each page, next to notification bell.
 */

import 'bootstrap/js/dist/dropdown';
import 'bootstrap/js/dist/button';
import { type ThemeValue } from '../utilities/theme';

const theme = defineModel<ThemeValue>('theme');

const THEMES = {
  auto: {
    icon: 'bi-circle-half',
    name: 'Auto'
  },
  light: {
    icon: 'bi-sun',
    name: 'Light mode'
  },
  dark: {
    icon: 'bi-moon-stars',
    name: 'Dark mode'
  }
} as const;
</script>

<template>
  <li class="nav-item dropdown">
    <button
      class="btn nav-link dropdown-toggle"
      href="#"
      type="button"
      data-bs-toggle="dropdown"
      aria-expanded="false"
      title="Change theme"
    >
      <i :class="`bi ${THEMES[theme]?.icon ?? 'bi-circle-half'}`"></i>
      <span class="d-md-none ms-1">Change theme</span>
    </button>
    <ul class="dropdown-menu dropdown-menu-end shadow">
      <li v-for="(data, key) in THEMES" :key="key">
        <button
          class="d-flex dropdown-item"
          :class="{ active: theme == key }"
          @click="() => (theme = key)"
        >
          <i :class="`me-1 bi ${data.icon}`"></i>{{ data.name }}
          <i class="opacity-75 ms-auto bi bi-check2" :hidden="theme != key"></i>
        </button>
      </li>
    </ul>
  </li>
</template>
