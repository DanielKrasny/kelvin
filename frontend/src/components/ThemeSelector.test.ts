import { describe, expect, it } from 'vitest';
import ThemeSelector from './ThemeSelector.vue';
import { mount } from '@vue/test-utils';
import { ThemeValue } from '../utilities/theme';

describe('ThemeSelector', () => {
    it('Auto option with no prop', () => {
        const wrapper = mount(ThemeSelector);
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Theme prop to auto', () => {
        const wrapper = mount(ThemeSelector, {
            props: {
                theme: 'auto' satisfies ThemeValue
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Theme prop to light', () => {
        const wrapper = mount(ThemeSelector, {
            props: {
                theme: 'light' satisfies ThemeValue
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Theme prop to dark', () => {
        const wrapper = mount(ThemeSelector, {
            props: {
                theme: 'dark' satisfies ThemeValue
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
});
