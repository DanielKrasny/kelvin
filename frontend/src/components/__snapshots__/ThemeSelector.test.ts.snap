// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeSelector > Auto option with no prop 1`] = `
"<li class="nav-item dropdown"><button class="btn nav-link dropdown-toggle" href="#" type="button" data-bs-toggle="dropdown" aria-expanded="false" title="Change theme"><i class="bi bi-circle-half"></i><span class="d-md-none ms-1">Change theme</span></button>
  <ul class="dropdown-menu dropdown-menu-end shadow">
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-circle-half"></i>Auto <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-sun"></i>Light mode <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-moon-stars"></i>Dark mode <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
  </ul>
</li>"
`;

exports[`ThemeSelector > Theme prop to auto 1`] = `
"<li class="nav-item dropdown"><button class="btn nav-link dropdown-toggle" href="#" type="button" data-bs-toggle="dropdown" aria-expanded="false" title="Change theme"><i class="bi bi-circle-half"></i><span class="d-md-none ms-1">Change theme</span></button>
  <ul class="dropdown-menu dropdown-menu-end shadow">
    <li><button class="d-flex dropdown-item active"><i class="me-1 bi bi-circle-half"></i>Auto <i class="opacity-75 ms-auto bi bi-check2"></i></button></li>
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-sun"></i>Light mode <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-moon-stars"></i>Dark mode <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
  </ul>
</li>"
`;

exports[`ThemeSelector > Theme prop to dark 1`] = `
"<li class="nav-item dropdown"><button class="btn nav-link dropdown-toggle" href="#" type="button" data-bs-toggle="dropdown" aria-expanded="false" title="Change theme"><i class="bi bi-moon-stars"></i><span class="d-md-none ms-1">Change theme</span></button>
  <ul class="dropdown-menu dropdown-menu-end shadow">
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-circle-half"></i>Auto <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-sun"></i>Light mode <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
    <li><button class="d-flex dropdown-item active"><i class="me-1 bi bi-moon-stars"></i>Dark mode <i class="opacity-75 ms-auto bi bi-check2"></i></button></li>
  </ul>
</li>"
`;

exports[`ThemeSelector > Theme prop to light 1`] = `
"<li class="nav-item dropdown"><button class="btn nav-link dropdown-toggle" href="#" type="button" data-bs-toggle="dropdown" aria-expanded="false" title="Change theme"><i class="bi bi-sun"></i><span class="d-md-none ms-1">Change theme</span></button>
  <ul class="dropdown-menu dropdown-menu-end shadow">
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-circle-half"></i>Auto <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
    <li><button class="d-flex dropdown-item active"><i class="me-1 bi bi-sun"></i>Light mode <i class="opacity-75 ms-auto bi bi-check2"></i></button></li>
    <li><button class="d-flex dropdown-item"><i class="me-1 bi bi-moon-stars"></i>Dark mode <i class="opacity-75 ms-auto bi bi-check2" hidden=""></i></button></li>
  </ul>
</li>"
`;
